### PR TITLE
use `git rev-list --stdin` instead of passing each remote ref

### PR DESCRIPTION
**Pull request by technoweenie**
Fixes https://github.com/github/git-lfs/issues/1318 by using `--stdin`. Hat tip to @alex-davidson for the idea. It doesn't work with the `--not` flag, so it now passes SHAs in with a `^` prefix.

### CODE REVIEW

File: lfs/scanner.go [hunk 1]
```diff
@@ -267,7 +267,7 @@ func ScanIndex() ([]*WrappedPointer, error) {
 
 // Get additional arguments needed to limit 'git rev-list' to just the changes in revTo
 // that are also not on remoteName.
-func revListArgsRefVsRemote(refTo, remoteName string) []string {
+func revListArgsRefVsRemote(refTo, remoteName string) ([]string, bool) {
```
Comment (ttaylorr) [inline lfs/scanner.go]: Can you document what the returned `bool` represents?

File: lfs/scanner.go [hunk 2]
```diff
@@ -297,25 +297,26 @@ func revListArgsRefVsRemote(refTo, remoteName string) []string {
...
+		commits := make([]string, 1, len(cachedRemoteRefs) + 1)
+		commits[0] = refTo
 		for _, cachedRef := range cachedRemoteRefs {
 			if !missingRefs.Contains(cachedRef.Name) {
-				ret = append(ret, fmt.Sprintf("refs/remotes/%v/%v", remoteName, cachedRef.Name))
+				commits = append(commits, "^"+cachedRef.Sha)
 			}
 		}
-		return ret
+		return commits
```
Comment (ttaylorr) [inline lfs/scanner.go]: Missing the second return here.

**Comment by ttaylorr** [comment]
LGTM other than 📝  comment! 👍

**Comment by technoweenie** [comment]
Welp, this clearly doesn't work. Tried different ways of passing the refs to `git rev-list` via stdin:

```diff
--- a/lfs/scanner.go
+++ b/lfs/scanner.go
@@ -304,7 +304,7 @@ func revListArgsRefVsRemote(refTo, remoteName string) ([]string, bool) {
                commits[0] = refTo
                for _, cachedRef := range cachedRemoteRefs {
                        if !missingRefs.Contains(cachedRef.Name) {
-                               commits = append(commits, "^"+cachedRef.Sha)
+                               commits = append(commits, fmt.Sprintf("^refs/remotes/%v/%v", remoteName, cachedRef.Name))
                        }
                }
                return commits, true
```

I made this slight tweak to the test to improve output:

```diff
diff --git a/lfs/scanner.go b/lfs/scanner.go
index b3ea4a4..ff0450f 100644
--- a/lfs/scanner.go
+++ b/lfs/scanner.go
@@ -455,6 +466,7 @@ begin_test "pre-push unfetched deleted remote branch & server GC"
   # all objects should now be there even though cached remote branch claimed it already had file3.dat
   for ((a=0; a < NUMFILES ; a++))
   do
+    echo "fetching object $a"
     assert_server_object "$reponame" "${oid[$a]}"
   done
```

Here's the error:

```
    fetching object 2
    *   Trying 127.0.0.1...
      % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                     Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Connected to 127.0.0.1 (127.0.0.1) port 55143 (#0)
    * Server auth using Basic with user 'user'
    > GET /test-pre-push-server-deleted-branch-gc.git/info/lfs/objects/ec3a53b52eb11808d59d29ffd01568edf2d62cdc9c02def84d0a11704d552de2 HTTP/1.1
    > Host: 127.0.0.1:55143
    > Authorization: Basic dXNlcjpwYXNz
    > User-Agent: curl/7.43.0
    > Accept: application/vnd.git-lfs+json
    >
    < HTTP/1.1 404 Not Found
```

In the test, [object 2](https://github.com/github/git-lfs/blob/c3b6a37558407679e03beaf5ef785a404ea99ddc/test/test-pre-push.sh#L418-L429) is the one pushed to the deleted branch. So, the client _should_ be detecting that it was deleted, and that it needs to re-upload the object.

**Comment by sinbad** [comment]
I may be being dumb here, but I can't see where you're actually passing the `--stdin` argument to `git rev-list`.

**Comment by technoweenie** [comment]
Lol yeah, that would do it. I was at gophercon the last 2 days, so just wrote these notes up until I could look at it today.

File: lfs/scanner.go [hunk 3]
```diff
@@ -267,7 +267,10 @@ func ScanIndex() ([]*WrappedPointer, error) {
 
 // Get additional arguments needed to limit 'git rev-list' to just the changes in revTo
 // that are also not on remoteName.
-func revListArgsRefVsRemote(refTo, remoteName string) []string {
+//
+// Returns a slice of string args, and false if --stdin is not needed.
```
Comment (ttaylorr) [inline lfs/scanner.go]: These docs need some ✨

**Pull request merged**